### DISCUSSION TITLE
Uppercase Keywords

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#1161] [General]
When I was fooling around with various Oberon-2 compilers, some enforced the use of uppercase keywords. Personally, I found that to be a bit of a PITA.

I have a `sed' command which works with a list of keywords that when run against a source code file will translate all lowercase keywords to uppercase. The file would have to be amended to include M3 keywords.

If anybody is interested I'll post the file and command.

**Comment by mikanystrom-intel** [comment]
The problem is, the uppercase keywords have a nice side-effect: you can use the lower-case versions as identifiers!

I always get annoyed when the C compiler rejects my personal way of writing "index i as floating point", i.e., "if"...

**Reply by dukester** [reply]
I suppose that you could still do that by appending an "_" to the lowercase version. Then when you're done, simply run the `sed' command and whammo! - all your keywords etc are in uppercase. I find it much easier to work in lowercase, than having to fiddle with the capsLock/shift keys.

To each his own though!

**Comment by dragisha** [comment]
There are lots of other languages with everything lowercase to choose from. People who do not like uppercase can survive, and comfortably, without!

**Comment by RodneyBates** [comment]
On 3/27/23 15:55, Duke Normandin wrote:
>
> I suppose that you could still do that by appending an "_" to the lowercase version. Then when you're done, simply run the `sed' command and whammo! - all your keywords etc are in uppercase. I find it much easier to work in lowercase, than having to fiddle with the capsLock/shift keys.
>
> To each his own though!
>

Decades ago, I came across an assembly of three foot-pedals shaped
like those of a piano, with switches on the pedals, intended
for an electronic piano or keyboard.  I wired them up to parallel
the shift, control and alt keys on a computer keyboard.  That
was a very fiddly job, with very little space for the cabling,
etc.  It made caps shifting, whether reserved words or identifiers
with caps, easier than stretching fingers or constantly hitting
caps-lock and loosing track of the current state.

For unknown reasons, I seem to have reverted mostly to using
the modifiers on the keyboard, but the pedals are still there
and get used occasionally.

Aside from the question of typing, I find all-caps reserved
words make it *far* easier to see the structure of code
from a more distant kind of gaze.  I understand the read-to-
write frequency ratio for code runs around 7-to-one, so
typing effort is a good time investment.

Back in the days of Algol-60 (anybody old enough?), they
used bold face in published journals, books etc. Computers
had caps only then.  After both letter cases became available,
I think Wirth may have been motivated to the all-caps idea
by this readability.  Not nearly as aesthetically pleasing,
to my taste, but gratifyingly easy to read.

**Reply by dukester** [reply]
I understand the points you make! And what a good idea your foot pedal keyboard enhancement. Cobbling a piece of electronics together is great fun to be sure. I remember building my first and only beat-frequency metal detector from a plan in "Popular Electronics". It was a hoot and it worked great. Thanks for your input!
